// docstate/lib/docstate.js
var stately = require("stately");

exports.control = function() {
    var me, safeMachine, safeStates = {}
        , cautiousMachine, unsafeStates = {}
        ;
    function makeMachines() {
        if (me._getType) safeStates._getType = me._getType;
        if (me._getState) unsafeStates._getState = me._getState;
        safeMachine = stately.define(safeStates);
        cautiousMachine = stately.define(unsafeStates);
    };
    
    function handle(doc) {
        safeMachine.handle(doc);
        cautiousMachine.handle(doc);
    };

    function registerSafeCallback(type, state, cb) {
        safeStates[type] = safeStates[type] || {};
        safeStates[type][state] = cb;
    }

    function registerUnsafeCallback(type, state, cb) {
        // todo this needs expiring lock
        unsafeStates[type] = unsafeStates[type] || {};
        unsafeStates[type][state] = cb;
    }
    me = {
        start : makeMachines,
        safe : registerSafeCallback,
        unsafe : registerUnsafeCallback,
        handle : handle
    }
    return me;
};
